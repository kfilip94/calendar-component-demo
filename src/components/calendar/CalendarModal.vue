<template>
  <div class="modalContainer">
    <div class="modaArrow" :style="arrowPosition"></div>
    <div class="modal">
      <div class="header">
        <button data-id="prevMonthBtn" class="navButton" @click="prevMonth">
          &lt;
        </button>
        <span class="monthLabel">{{ monthName }}</span>
        <button data-id="nextMonthBtn" class="navButton" @click="nextMonth">
          &gt;
        </button>
      </div>
      <div class="monthGrid">
        <div v-for="dayName in dayNames" :key="dayName" class="dayName">
          {{ dayName }}
        </div>
        <CalendarDayItem
          v-for="day in days"
          :key="day.key"
          :day="day"
          :startDate="startDate"
          :endDate="endDate"
          :lastAvailableEndDate="lastAvailableEndDate"
          :todayDate="todayDate"
          :editMode="editMode"
          @onDayClick="editReservationDays"
        />
      </div>
    </div>
  </div>
</template>

<script>
import {
  getCalendarDays,
  getPreviousMonthDate,
  getNextMonthDate,
  getTodayDate,
  EDIT_MODES,
  editModeValidator,
} from "./CalendarUtils";
import CalendarDayItem from "./CalendarDayItem.vue";

const { CHECK_IN, CHECK_OUT } = EDIT_MODES;

export default {
  name: "CalendarGrid",
  components: { CalendarDayItem },
  props: {
    startDate: Number,
    endDate: Number,
    editMode: {
      type: String,
      validator: editModeValidator,
    },
    availableDates: Array,
  },

  data() {
    return {
      todayDate: getTodayDate(),
      selectedDate: new Date(),
      dayNames: ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
      days: [],
    };
  },

  mounted() {
    this.updateCalendarDays();
  },

  computed: {
    lastAvailableEndDate() {
      const { startDate, availableDates, getLastAvailableEndDate } = this;
      return startDate
        ? getLastAvailableEndDate(startDate, availableDates)
        : null;
    },

    monthName() {
      const month = this.selectedDate.toLocaleString("en-US", {
        month: "long",
      });
      const year = this.selectedDate.getFullYear();
      return `${month} ${year}`;
    },

    arrowPosition() {
      const arrowSide = this.editMode === CHECK_IN ? "left" : "right";
      return { [arrowSide]: "32px" };
    },
  },

  watch: {
    selectedDate() {
      this.updateCalendarDays();
    },
    startDate() {},
  },

  methods: {
    updateCalendarDays() {
      this.days = getCalendarDays(this.selectedDate, this.availableDates);
    },

    prevMonth() {
      this.selectedDate = getPreviousMonthDate(this.selectedDate);
    },

    nextMonth() {
      this.selectedDate = getNextMonthDate(this.selectedDate);
    },

    getLastAvailableEndDate(startDate, availableDates) {
      const availableEndDates = availableDates.find(
        ({ from, to }) => from <= startDate && startDate <= to
      );
      return availableEndDates ? availableEndDates.to : null;
    },

    editReservationDays(selectedDay) {
      let startDate = this.startDate;
      let endDate = this.endDate;
      let editMode = this.editMode;

      if (this.editMode === CHECK_IN) {
        startDate = selectedDay.miliseconds;
        editMode = CHECK_OUT;

        const lastAvailableEndDate = this.getLastAvailableEndDate(
          startDate,
          this.availableDates
        );

        //reset endDate if needed
        if (
          endDate &&
          (startDate > endDate || endDate > lastAvailableEndDate)
        ) {
          endDate = null;
        }
      } else if (this.editMode === CHECK_OUT) {
        endDate = selectedDay.miliseconds;
        editMode = null;
      }

      this.$emit("on-reservation-change", { startDate, endDate, editMode });
    },
  },
};
</script>

<style lang="scss" scoped>
@import "./styles.scss";

.modalContainer {
  position: relative;
  width: 100%;

  .modaArrow {
    box-shadow: 0px 0px 4px 0px rgba(122, 122, 122, 0.5);
    display: block;
    width: 12px;
    height: 12px;
    position: absolute;
    background: white;
    border: $border;
    transform: rotate(45deg);
    top: 1px;
  }

  .modal {
    background-color: white;
    top: 8px;
    border-radius: 2px;
    position: absolute;
    width: 100%;
    overflow: hidden;
    box-shadow: 0px 0px 4px 0px rgba(122, 122, 122, 0.5);
  }

  .monthGrid {
    background-color: white;
    box-sizing: border-box;
    color: grey;
    display: grid;
    font-size: $font-size;
    grid-template-columns: repeat(7, 1fr);
    padding: $padding;
    width: 100%;
  }

  .header {
    background-color: $theme-color;
    color: white;
    font-style: bold;
    padding: $padding;
    display: flex;
    justify-content: space-between;
    font-weight: $font-weight-bold;
  }

  .navButton {
    background: transparent;
    border: 0;
    font-weight: $font-weight-bold;
    cursor: pointer;
    text-decoration: underline;
    outline: none;
  }

  .monthLabel {
    text-transform: capitalize;
  }
}
</style>