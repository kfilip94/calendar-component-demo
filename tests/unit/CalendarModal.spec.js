import { mount } from '@vue/test-utils'
import CalendarModal from '@/components/calendar/CalendarModal.vue';

describe('CalendarModal.vue', () => {
  const lastAvailableDate = Date.parse(new Date('2020-12-28'));
  const availableDates = [
    {
      from: Date.parse(new Date('2020-11-4')),
      to: Date.parse(new Date('2020-12-3'))
    }, {
      from: Date.parse(new Date('2020-12-12')),
      to: Date.parse(new Date('2020-12-16'))
    }, {
      from: Date.parse(new Date('2020-12-18')),
      to: lastAvailableDate
    }, {
      from: Date.parse(new Date('2020-12-30')),
      to: Date.parse(new Date('2021-1-24'))
    }
  ];
  
  const wrapper = mount(CalendarModal, {
    props: {
      selectedDate: new Date('2020-12-1'),
      availableDates
    }
  })

  it('select next month', async () => {
    await wrapper.find('[data-id="prevMonthBtn"]').trigger('click');
    expect(wrapper.element).toMatchSnapshot()
  })

  it('select previous month', async () => {
    await wrapper.setProps({
      selectedDate: new Date('2021-2-1')
    })
    await wrapper.find('[data-id="nextMonthBtn"]').trigger('click');
    expect(wrapper.element).toMatchSnapshot()
  })
});