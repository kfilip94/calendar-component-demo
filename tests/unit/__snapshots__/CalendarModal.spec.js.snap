// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarModal.vue select next month 1`] = `
<div
  class="modalContainer"
  selecteddate="Tue Dec 01 2020 00:00:00 GMT+0100 (Central European Standard Time)"
>
  <div
    class="modaArrow"
    style="right: 32px;"
  />
  <div
    class="modal"
  >
    <div
      class="header"
    >
      <button
        class="navButton"
        data-id="prevMonthBtn"
      >
        &lt;
      </button>
      <span
        class="monthLabel"
      >
        November 2020
      </span>
      <button
        class="navButton"
        data-id="nextMonthBtn"
      >
        &gt;
      </button>
    </div>
    <div
      class="monthGrid"
    >
      
      <div
        class="dayName"
      >
        Sun
      </div>
      <div
        class="dayName"
      >
        Mon
      </div>
      <div
        class="dayName"
      >
        Tue
      </div>
      <div
        class="dayName"
      >
        Wed
      </div>
      <div
        class="dayName"
      >
        Thu
      </div>
      <div
        class="dayName"
      >
        Fri
      </div>
      <div
        class="dayName"
      >
        Sat
      </div>
      
      
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          25
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          26
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          27
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          28
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          29
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          30
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          31
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          1
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          2
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          3
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          4
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          5
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          6
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          7
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          8
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          9
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          10
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          11
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          12
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          13
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          14
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          15
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          16
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          17
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          18
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          19
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          20
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          21
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          22
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          23
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          24
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          25
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          26
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          27
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          28
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          29
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          30
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          1
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          2
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          3
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          4
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          5
        </span>
      </button>
      
    </div>
  </div>
</div>
`;

exports[`CalendarModal.vue select previous month 1`] = `
<div
  class="modalContainer"
  selecteddate="Mon Feb 01 2021 00:00:00 GMT+0100 (Central European Standard Time)"
>
  <div
    class="modaArrow"
    style="right: 32px;"
  />
  <div
    class="modal"
  >
    <div
      class="header"
    >
      <button
        class="navButton"
        data-id="prevMonthBtn"
      >
        &lt;
      </button>
      <span
        class="monthLabel"
      >
        December 2020
      </span>
      <button
        class="navButton"
        data-id="nextMonthBtn"
      >
        &gt;
      </button>
    </div>
    <div
      class="monthGrid"
    >
      
      <div
        class="dayName"
      >
        Sun
      </div>
      <div
        class="dayName"
      >
        Mon
      </div>
      <div
        class="dayName"
      >
        Tue
      </div>
      <div
        class="dayName"
      >
        Wed
      </div>
      <div
        class="dayName"
      >
        Thu
      </div>
      <div
        class="dayName"
      >
        Fri
      </div>
      <div
        class="dayName"
      >
        Sat
      </div>
      
      
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          29
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          30
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          1
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          2
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          3
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          4
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          5
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          6
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          7
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          8
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          9
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          10
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          11
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          12
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          13
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          14
        </span>
      </button>
      <button
        class="today day"
      >
        <span
          class="dayNumber"
        >
          15
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          16
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          17
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          18
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          19
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          20
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          21
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          22
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          23
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          24
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          25
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          26
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          27
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          28
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          29
        </span>
      </button>
      <button
        class="unavailableDay day"
      >
        <span
          class="dayNumber"
        >
          30
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          31
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          1
        </span>
      </button>
      <button
        class="day"
      >
        <span
          class="dayNumber"
        >
          2
        </span>
      </button>
      
    </div>
  </div>
</div>
`;
